journalbeat/input/input.go

```diff
@@ -121,43 +121,48 @@ func (i *Input) Run() {
 	}
 	defer client.Close()
 
-	for {
-		select {
-		case <-i.done:
-			return
-		default:
-			i.publishAll(client)
-		}
-	}
-
+	i.publishAll(client)
 }
 
 func (i *Input) publishAll(client beat.Client) {
 	out := make(chan *beat.Event)
-	var wg sync.WaitGroup
 
+	var wg sync.WaitGroup
 	merge := func(in chan *beat.Event) {
 		wg.Add(1)
 
 		go func(c chan *beat.Event) {
-			for v := range c {
-				out <- v
+			defer wg.Done()
+			for {
+				select {
+				case <-i.done:
+					return
+				case v := <-c:
+					out <- v
+				}
 			}
-			wg.Done()
 		}(in)
 	}
+
+	// close output channel after all input channels are merged or beats is stopped
 	go func() {
 		wg.Wait()
 		close(out)
 	}()
```

Comment:
@kvch There is no need to create a goroutine to wait on the _WaitGroup_, instead it could be done in a defer.

Reply:
If I close using `defer` journalbeat panics with "send on a closed channel". I am not sure how exactly your solution supposed to work. Could you give more context? 
I mean OFC I understand `defer` is called at the end of the function. But how would you rewrite the code so it could work?

Reply:
Using `defer` sometimes works correctly. But most of the time journalbeat panics, either because of a null pointer dereference or sending on a closed channel. So closing the channel after wait is necessary to provide stable behaviour.